### PR TITLE
Recursive CTE data race fix

### PR DESCRIPTION
**Pull request by kitaisreal**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit d237133ec00cf51e3eca7edb06287a6cc751c743 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/63125/d237133ec00cf51e3eca7edb06287a6cc751c743/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Review by alexey-milovidov — APPROVED** [review]

File: src/Interpreters/executeQuery.cpp [hunk 1]
```diff
@@ -259,15 +259,20 @@ addStatusInfoToQueryLogElement(QueryLogElement & element, const QueryStatusInfo
 
     /// We need to refresh the access info since dependent views might have added extra information, either during
     /// creation of the view (PushingToViews chain) or while executing its internal SELECT
-    const auto & access_info = context_ptr->getQueryAccessInfo();
+
+    /// We copy QueryAccessInfo for thread-safety, because it is possible that query context can be modified by some processor even
+    /// after query is finished
+    ContextData::QueryAccessInfo access_info(context_ptr->getQueryAccessInfo());
```
Comment (Algunenano) [inline src/Interpreters/executeQuery.cpp:265]: I'm having a look at whether it'd be better to mark the members as `TSA_GUARDED_BY(mutex)` and force acquiring the mutex here.

**Review by Algunenano — APPROVED** [review]
Thanks a lot for the quick fix @kitaisreal 

I've made 2 small changes:
* Use TSA annotations which force acquiring the mutex in several places (correctly), so we don't need to copy the object. I don't expect there to be any mutex contention since logging should not be happening with active pipelines (but tsan will still complain if we don't).
* Some changes in the test to avoid flakiness (session_timezone) and my bad environment (enable_global_with_statement).

Let's confirm everything is green now and merge

**Comment by kitaisreal** [comment]
> Thanks a lot for the quick fix @kitaisreal
> 
> I've made 2 small changes:
> 
> * Use TSA annotations which force acquiring the mutex in several places (correctly), so we don't need to copy the object. I don't expect there to be any mutex contention since logging should not be happening with active pipelines (but tsan will still complain if we don't).
> * Some changes in the test to avoid flakiness (session_timezone) and my bad environment (enable_global_with_statement).
> 
> Let's confirm everything is green now and merge

Thanks. Let's wait for CI checks and merge.

**Pull request merged**